mono/metadata/loader.c

```diff
@@ -1568,23 +1569,85 @@ mono_lookup_pinvoke_call (MonoMethod *method, const char **exc_class, const char
 		}
 
 		if (!module && !is_absolute) {
-			void *iter = NULL;
-			char *mdirname = g_path_get_dirname (image->name);
-			while ((full_name = mono_dl_build_path (mdirname, file_name, &iter))) {
-				module = cached_module_load (full_name, MONO_DL_LAZY, &error_msg);
-				if (!module) {
-					mono_trace (G_LOG_LEVEL_INFO, MONO_TRACE_DLLIMPORT,
-						"DllImport error loading library '%s': '%s'.",
-								full_name, error_msg);
-					g_free (error_msg);
-				} else {
-					found_name = g_strdup (full_name);
+			void *iter;
+			char *mdirname;
+
+			for (j = 0; j < 3; ++j) {
+				iter = NULL;
+				mdirname = NULL;
+				switch (j) {
+					case 0:
+						mdirname = g_path_get_dirname (image->name);
+						break;
+					case 1: /* @executable_path@/../lib */
```

Comment:
Both case 1 and 2 have pretty much the same code so I'd say you could factor it into a function. And then could also probably get rid of the switch.